File: problemset/594-LongestHarmoniousSubsequence/lhs.py
class Solution:
    def findLHS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        self.numminplace ={}
        self.nummaxplace ={}
        for i in range(len(nums)):
            if nums[i] in self.numminplace:
                self.nummaxplace[nums[i]][1] += 1
            else:
                self.numminplace[nums[i]] = [i, 1]
                self.nummaxplace[nums[i]] = [i, 1]
        maxlength = 0
        sort = sorted(nums)
        for i in range(len(sort)-1):
            if sort[i+1]-1 == sort[i]:
                maxplace = self.nummaxplace[sort[i]][1] + self.nummaxplace[sort[i+1]][1]
                maxlength = maxplace if maxplace>maxlength else maxlength
        return maxlength

File: problemset/594-LongestHarmoniousSubsequence/lhs.go
package lhs

import (
	"sort"
)

type numRecord struct {
	index          int
	occurencecount int
}

func maxInt(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func minInt(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func findLHS(nums []int) int {
	sorted := make([]int, len(nums))
	numminplace := make(map[int]*numRecord, 0)
	nummaxplace := make(map[int]*numRecord, 0)
	copy(sorted[:], nums)
	sort.Ints(sorted)
	for index, val := range nums {
		if _, ok := numminplace[val]; !ok {
			numminplace[val] = new(numRecord)
			numminplace[val].index = index
			numminplace[val].occurencecount = 1
			nummaxplace[val] = new(numRecord)
			nummaxplace[val].index = index
			nummaxplace[val].occurencecount = 1
			continue
		}
		nummaxplace[val] = &numRecord{
			index:          index,
			occurencecount: nummaxplace[val].occurencecount + 1,
		}
	}
	maxlength := 0
	for i := 0; i < len(sorted)-1; i++ {
		if sorted[i+1]-1 == sorted[i] {
			maxplace := nummaxplace[sorted[i]].occurencecount + nummaxplace[sorted[i+1]].occurencecount
			maxlength = maxInt(maxlength, maxplace)
		}
	}
	return maxlength
}
